server/console_api_explorer.go

```diff
@@ -0,0 +1,268 @@
+	if r.method.Type.In(2) == reflect.TypeOf(&empty.Empty{}) {
+		if in.Body != "" {
+			s.logger.Error("Body passed to an api call that doesn't accept any.", zap.String("method", in.Method))
+			return nil, status.Error(codes.InvalidArgument, "Api method doesn't accept a request body.")
+		}
+		args[2] = reflect.ValueOf(&empty.Empty{})
+	} else {
+		request := reflect.New(r.request).Interface().(proto.Message)
+		err = jsonpb.UnmarshalString(in.Body, request)
+		if err != nil {
+			s.logger.Error("Error parsing method request body.", zap.String("method", in.Method), zap.Error(err))
+			return nil, status.Error(codes.InvalidArgument, "Error parsing method request body.")
+		}
+		args[2] = reflect.ValueOf(request)
+	}
+	out := r.method.Func.Call(args)
+	cval := out[0].Interface()
+	cerr := out[1].Interface()
+	if cerr != nil {
+		return &console.CallApiEndpointResponse{
+			Body:         "",
+			ErrorMessage: cerr.(error).Error(),
+		}, nil
+	} else {
+		var j string
+		if cval != nil {
+			m := new(jsonpb.Marshaler)
+			j, err = m.MarshalToString(cval.(proto.Message))
+			if err != nil {
+				s.logger.Error("Error serializing method response body.", zap.String("method", in.Method), zap.Error(err))
+				return nil, status.Error(codes.Internal, "Error serializing method response body.")
+			}
+		}
+		return &console.CallApiEndpointResponse{
+			Body: j,
+		}, nil
+	}
+
+}
+
+func (s *ConsoleServer) extractApiCallContext(ctx context.Context, in *console.CallApiEndpointRequest, userIdOptional bool) (context.Context, error) {
+	var callCtx context.Context
+	if strings.HasPrefix(in.Method, "Authenticate") {
+		callCtx = ctx
+	} else if in.UserId == "" {
+		if !userIdOptional {
+			s.logger.Error("Error calling a built-in RPC function without a user_id.", zap.String("method", in.Method))
+			return nil, status.Error(codes.InvalidArgument, "Built-in RPC functions require a user_id.")
+		} else {
+			callCtx = ctx
+		}
+	} else {
+		row := s.db.QueryRowContext(ctx, "SELECT username FROM users WHERE id = $1", in.UserId)
+		var dbUsername string
+		userUUID, err := uuid.FromString(in.UserId)
+		if err != nil {
+			s.logger.Error("Invalid user uuid.", zap.String("method", in.Method))
+			return nil, status.Error(codes.InvalidArgument, "Invalid user uuid.")
+		}
+		err = row.Scan(&dbUsername)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				s.logger.Error("User id not found.", zap.String("method", in.Method))
+				return nil, status.Error(codes.InvalidArgument, "User id not found.")
+			}
+			s.logger.Error("Error looking up user account.", zap.String("method", in.Method), zap.Error(err))
+			return nil, status.Error(codes.Internal, "Error looking up user account.")
+		}
+		callCtx = context.WithValue(context.WithValue(ctx, ctxUserIDKey{}, userUUID), ctxUsernameKey{}, dbUsername)
+	}
+	return callCtx, nil
+
+}
+
+func (s *ConsoleServer) ListApiEndpoints(ctx context.Context, _ *empty.Empty) (*console.ApiEndpointList, error) {
+	if s.methodListCache == nil {
```

Comment:
This whole block is a problem for concurrency. Instead we should initialise this cache once in the server setup phase.

Reply:
Firstly @mofirouz wanted it initialized on the first call and yes, double-checking would be more appropriate here but it is a rare operation - initializing on the startup will of course solve it.

Reply:
I think there is a little bit of confusion. I only wanted storage collections to be cached on first use. This is safe to cache on server start.